src/compiler/checker.ts

```diff
@@ -3643,9 +3630,17 @@ module ts {
         }
 
         function getReducedTypeOfUnionType(type: UnionType): Type {
-            // If union type was created without subtype reduction, perform the deferred reduction now
+            // If union type was created without subtype reduction, perform the deferred reduction now. If a circularity
+            // is detected, simply use the type itself.
             if (!type.reducedType) {
-                type.reducedType = getUnionType(type.types, /*noSubtypeReduction*/ false);
+                type.reducedType = resolvingType;

```

Comment:
The kind of cycles detected by the type resolution stack has nothing to do with reduced union types and would never involve union type reduction. This is a distinct and unrelated kind of recursion. It so happens to use a marker type called `resolvingType` for cycle detection, but that's just a name. Maybe it should be renamed `markerType` or `circularType` since it isn't actually used by type resolution anymore (we have the resolution stack for that).